### ISSUE TITLE
Cargo `vendor` doesn't include dependency's hidden files

### ISSUE THREAD

**Issue by joshuachp** [rust-lang/cargo#13662]
### Problem

Vendoring a dependency ignores hidden files present in the published crate and that are required to build it.

### Steps

1. Create a library dependency or crate that `include_str!(".hidden")` (e.g. [test-dep-cargo-hidden](https://github.com/joshuachp/test-dep-cargo-hidden))
2. This crate need to be published on a registry, the crate will include the `.hidden` file
3. Create another crate that uses the published dependency
4. Use `cargo vendor` and in the vendor directory the `.hidden` file is missing
5. Building with the vendor will fail, while normal builds will succeed

### Possible Solution(s)

Vendoring should probably include all the files of the published package.

I believe those line caused this discrepancy:

 https://github.com/rust-lang/cargo/blob/3b43aa875fa262f53bdcc5ad4c0f677802630186/src/cargo/sources/path.rs#L159-L162

This will cause the `publish` command to list all the files in the git repo of the dependency, while the `vendor` will use the sparse cache, which is not a git repository, to copy the files.

### Notes

An example of this is having a dependency using `sqlx` offline files that are stored in the `.sqlx/` directory.

Relevant documentation, issues and PRs I found:
- https://doc.rust-lang.org/cargo/reference/manifest.html#the-exclude-and-include-fields
- https://github.com/rust-lang/cargo/issues/7183
- https://github.com/rust-lang/cargo/pull/7680

### Version

```text
cargo 1.77.0 (3fe68eabf 2024-02-29)
release: 1.77.0
commit-hash: 3fe68eabf93cbf3772bbcad09a9206c783e2de3f
commit-date: 2024-02-29
host: x86_64-unknown-linux-gnu
libgit2: 1.7.2 (sys:0.18.2 vendored)
libcurl: 8.5.0-DEV (sys:0.4.70+curl-8.5.0 vendored ssl:OpenSSL/1.1.1w)
ssl: OpenSSL 1.1.1w  11 Sep 2023
os: NixOS 24.5.0 [64-bit]
```

**Comment by charmitro** [comment]
Does `cargo vendor` include dotfiles (`.file`) when specified in the `package.include` field? 

Given that dotfiles sometimes contain sensitive information, it would be prudent to include only those explicitly mentioned in the `package.include` field.

**Comment by joshuachp** [comment]
Yes it does, but I am talking specifically of a package already published on a registry. The published crate already includes the `.files` since they were added in the VCS and not in the `excluede`

**Comment by charmitro** [comment]
> Yes it does, but I am talking specifically of a package already published on a registry. The published crate already includes the `.files` since they were added in the VCS and not in the `exclude` list, but the vendor doesn't.

Ah, now I get it. We should be including dotfiles just like `publish` does, except when they are present in the `package.exclude` field.

**Comment by epage** [comment]
When we `cargo publish`, we extract the source from the git repo and make it independent.  `cargo vendor` then applied that same logic to the `.crate`.  Because it was no longer in a git repo, we fallback to our other inclusion rules which is to ignore all hidden files.

#15514 bypasses the second step and directly extracts a `.crate` to the vendor directory, so this should now be resolved and  I'm going to close this.  If there is something I missed, let us know and we can re-open it.

FYI there is a discussion on [Internals](https://internals.rust-lang.org/t/exclude-github-and-gitignore-from-published-packages-by-default/23700/11) about, through an edition, changing the implicit inclusion rules for git to also ignore hidden files.  Might be good to note the use case for this hidden file on that to help raise awareness of a practical place people do this so we can weigh the impact of such a change to the implicit inclusion.
